engine/render/src/render/render.cpp

```diff
@@ -681,6 +679,116 @@ namespace dmRender
+        if (unit >= render_context->m_TextureBindTable.Size())
+        {
+            render_context->m_TextureBindTable.SetCapacity(unit + 1);
+
+            // Make sure new data area is zeroed out
+            uint32_t fill_index_start = render_context->m_TextureBindTable.Size();
+            uint32_t fill_size = render_context->m_TextureBindTable.Remaining() * sizeof(TextureBinding);
+
+            render_context->m_TextureBindTable.SetSize(render_context->m_TextureBindTable.Capacity());
+            memset(&render_context->m_TextureBindTable[fill_index_start], 0, fill_size);
+        }
+
+        render_context->m_TextureBindTable[unit].m_Texture     = texture;
+        render_context->m_TextureBindTable[unit].m_Samplerhash = 0;
+    }
+
+    static void SetRenderContextTextures(HRenderContext render_context, HMaterial material, dmGraphics::HTexture* textures)
+    {
+        uint32_t num_bindings = render_context->m_TextureBindTable.Size();
+        uint32_t last_zero_index = -1;
+        for (uint32_t i = 0; i < num_bindings; ++i)
+        {
+            if (render_context->m_TextureBindTable[i].m_Samplerhash)
+            {
+                int32_t sampler_index = GetMaterialSamplerIndex(material, render_context->m_TextureBindTable[i].m_Samplerhash);
+                if (sampler_index >= 0 && sampler_index < RenderObject::MAX_TEXTURE_COUNT)
+                {
+                    if (sampler_index < RenderObject::MAX_TEXTURE_COUNT)
+                    {
+                        textures[sampler_index] = render_context->m_TextureBindTable[i].m_Texture;
+                    }
+                    else
+                    {
+                        dmLogOnceWarning("Unable to bind texture '%s' to unit %d, max %d texture units are supported.",
+                            dmHashReverseSafe64(render_context->m_TextureBindTable[i].m_Samplerhash),
+                            sampler_index, RenderObject::MAX_TEXTURE_COUNT);
+                    }
+                }
+            }
+            else if (textures[i] == 0)
+            {
+                if (i < RenderObject::MAX_TEXTURE_COUNT)
+                {
+                    textures[i] = render_context->m_TextureBindTable[i].m_Texture;
+                }
+                else
+                {
+                    dmLogOnceWarning("Unable to bind texture to unit %d, max %d texture units are supported.",
+                        i, RenderObject::MAX_TEXTURE_COUNT);
+                }
+            }
+
+            if (render_context->m_TextureBindTable[i].m_Texture == 0)
+            {
+                if (last_zero_index == -1)
+                {
+                    last_zero_index = i;
+                }
+            }
+            else
+            {
+                last_zero_index = -1;
+            }
+        }
+
+        // Trim the iteration space
+        if (last_zero_index != -1)
+        {
+            render_context->m_TextureBindTable.SetSize(last_zero_index);
+        }
```

Comment:
Do we want to have a separate function for trimming it?
I think the function name doesn't really reflect that it does two things: getting textures, and also modifying the current state.

Reply:
Yeah, I did it here since we are already looping over the textures, but for clarity I guess we could have a separate trim function